### DISCUSSION TITLE
Monitor Spanning by Choosing Monitors

### DISCUSSION THREAD

**Discussion by murtaza-nasir** [projectM-visualizer/projectm#627] [Ideas]
Hi,

First of all, thank you for the amazing work! I was wondering if it would be possible to have the monitor span function extended to be able to allow to choose which monitors to span to. That way I can have my music player on one of my monitors and the rest can be filled with the visualization. Also, would it be possible to have the option to switch from stretching the image across all monitors to having each monitor have its own (the same or independent) render of the same visualization.

Thanks

**Comment by kblaschke** [comment]
Multi-monitor stuff is quite involved, but possible generally. Stetching is surely the easiest part, but can be limited by the maximum texture size the GPU allows (8K for most desktop GPUs), so spanning the same image over 3 4K monitors is hardly possible.

In addition to that, monitors can have any arrangement in relation to each other, so the resulting desktop isn't necessarily the same area visible on the monitors. There can be empty areas, or, depending on the OS, possibly even gaps between them. This would cut out parts of the image, which isn't desirable.

The probably best-looking solution would be displaying an individual visualization on each monitor. This will result in increased CPU/GPU utilization as it basically requires to render a frame once per monitor. This could be done in parallel and make use of multi-core CPUs, but it will probably render slowly on dated or cheap hardware. Displaying a copy of a frame would mitigate the performance impact, but if the monitors have different resolutions, scaling must be applied (ideally render on the largest res and scale down), which can reduce the image quality on non-native copies of the image.

One way to achieve it without changes would be running multiple instances of projectMSDL in parallel, one on each monitor. Will make it impossible to sync the preset transitions etc., but should work perfectly fine. The [new SDL2 frontend](https://github.com/projectM-visualizer/frontend-sdl2) has an option `window.monitor`, which you can set to the display index the window should show up on. You can also provide it via command line using the `/monitor` (Windows) or `--monitor` (Linux/macOS) parameter. Combined with fullscreen, it will open the window exactly where you want it to be.
